File: model/layers/tick_loop.py
"""
model/layers/tick_loop.py

Iterative tick loop: attend → update neurons → recompute sync → check confidence.

The loop terminates when confidence ≥ threshold OR max_ticks is reached.
"""

from __future__ import annotations

from dataclasses import dataclass

import torch

from .correlation import pearson_correlation
from .sync_attention import SyncAttention
from .confidence_head import ConfidenceHead


MAX_TICKS = 32
DEFAULT_CONFIDENCE_THRESHOLD = 0.9


@dataclass
class TickLoopLog:
    ticks_run: int
    max_ticks: int
    converged: bool
    confidence: float
    confidence_threshold: float


class TickLoop(torch.nn.Module):
    """Runs the iterative tick loop over a set of neuron embeddings.

    Parameters
    ----------
    attention : SyncAttention
    confidence_head : ConfidenceHead
    max_ticks : int
    confidence_threshold : float
    """

    def __init__(
        self,
        attention: SyncAttention,
        confidence_head: ConfidenceHead,
        max_ticks: int = MAX_TICKS,
        confidence_threshold: float = DEFAULT_CONFIDENCE_THRESHOLD,
    ) -> None:
        super().__init__()
        self.attention = attention
        self.confidence_head = confidence_head
        self.max_ticks = max_ticks
        self.confidence_threshold = confidence_threshold

    def forward(
        self,
        embeddings: torch.Tensor,
    ) -> tuple[torch.Tensor, torch.Tensor, TickLoopLog]:
        """Run tick loop starting from initial embeddings.

        Parameters
        ----------
        embeddings : torch.Tensor
            Shape (N, embed_dim).

        Returns
        -------
        context : torch.Tensor
            Shape (N, value_dim) — final attended context after loop terminates.
        sync_matrix : torch.Tensor
            Shape (N, N) — final pairwise sync matrix.
        log : TickLoopLog
            Structured log of loop execution.
        """
        n = embeddings.shape[0]
        context = embeddings  # initial context = embeddings

        # Initial sync matrix
        sync = _compute_sync(context)

        confidence = torch.tensor(0.0)
        ticks_run = 0
        converged = False

        for tick in range(self.max_ticks):
            ticks_run = tick + 1

            # Attend: query from sync, K/V from current context
            context = self.attention(sync, context)

            # Recompute sync from updated context
            sync = _compute_sync(context)

            # Check confidence
            confidence = self.confidence_head(sync)
            if confidence.item() >= self.confidence_threshold:
                converged = True
                break

        log = TickLoopLog(
            ticks_run=ticks_run,
            max_ticks=self.max_ticks,
            converged=converged,
            confidence=float(confidence.item()),
            confidence_threshold=self.confidence_threshold,
        )
        return context, sync, log


def _compute_sync(context: torch.Tensor) -> torch.Tensor:
    """Build N×N Pearson correlation sync matrix from context vectors."""
    n = context.shape[0]
    sync = torch.zeros(n, n, dtype=context.dtype, device=context.device)
    for i in range(n):
        for j in range(n):
            sync[i, j] = pearson_correlation(context[i], context[j])
    return sync

File: src/lib/model/tick-loop.ts
import { SyncAttention } from "./sync-attention";
import { ConfidenceHead } from "./confidence-head";
import { pearsonCorrelation } from "./correlation";

// ── Constants ─────────────────────────────────────────────────────────────────

export const MAX_TICKS = 32;
export const DEFAULT_CONFIDENCE_THRESHOLD = 0.9;

// ── Types ─────────────────────────────────────────────────────────────────────

export interface TickLoopConfig {
  maxTicks?: number;
  confidenceThreshold?: number;
}

export interface TickLoopResult {
  /** Final context vector produced by the last attention step */
  context: number[];
  /** Final confidence score [0, 1] */
  confidence: number;
  /** Number of ticks actually executed */
  ticksRun: number;
  /** true if stopped because confidence exceeded threshold, false if maxTicks reached */
  converged: boolean;
  /** Structured log entry — suitable for appending to an audit trail or metrics system */
  log: TickLoopLog;
}

export interface TickLoopLog {
  ticksRun: number;
  maxTicks: number;
  converged: boolean;
  confidence: number;
  confidenceThreshold: number;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/**
 * Compute the N×N pairwise Pearson correlation matrix for a set of neuron
 * activation sequences. neuronActivations[i] is the activation history of
 * neuron i (length = ticks so far). Returns the identity matrix when fewer
 * than 2 ticks have elapsed (correlation undefined).
 */
function computeSyncMatrix(neuronActivations: number[][]): number[][] {
  const n = neuronActivations.length;
  return Array.from({ length: n }, (_, i) =>
    Array.from({ length: n }, (_, j) => {
      if (i === j) return 1;
      // Need at least 2 timesteps for Pearson correlation
      if (neuronActivations[i].length < 2) return 0;
      return pearsonCorrelation(neuronActivations[i], neuronActivations[j]);
    }),
  );
}

/**
 * Update neuron activations given the new context vector.
 * Each neuron i takes the i-th component of the context (mod valueDim)
 * and appends it to its history.
 */
function updateNeurons(
  neuronActivations: number[][],
  context: number[],
): void {
  const n = neuronActivations.length;
  for (let i = 0; i < n; i++) {
    neuronActivations[i].push(context[i % context.length]);
  }
}

// ── Main function ─────────────────────────────────────────────────────────────

/**
 * TickLoop — the Piro model's inference loop.
 *
 * Each tick:
 *   1. Attend — SyncAttention produces a context vector from the sync matrix + embeddings
 *   2. Update neurons — append context components to neuron activation histories
 *   3. Recompute sync matrix — Pearson correlations over the updated activation histories
 *   4. Check confidence — ConfidenceHead scores the new sync matrix; stop if above threshold
 *
 * Stops early when confidence > threshold, or after maxTicks iterations.
 *
 * @param attention    Initialised SyncAttention layer
 * @param confHead     Initialised ConfidenceHead
 * @param embeddings   Input token embeddings (seqLen × embedDim)
 * @param nNeurons     Number of neurons in the network
 * @param config       Optional: maxTicks, confidenceThreshold
 */
export function tickLoop(
  attention: SyncAttention,
  confHead: ConfidenceHead,
  embeddings: number[][],
  nNeurons: number,
  config: TickLoopConfig = {},
): TickLoopResult {
  const maxTicks = config.maxTicks ?? MAX_TICKS;
  const threshold = config.confidenceThreshold ?? DEFAULT_CONFIDENCE_THRESHOLD;

  // Each neuron starts with an empty activation history
  const neuronActivations: number[][] = Array.from({ length: nNeurons }, () => []);

  // Seed the sync matrix as identity (diagonal = 1, off-diagonal = 0)
  // before we have enough history for Pearson correlation
  let syncMatrix = computeSyncMatrix(neuronActivations); // starts as identity (no history)

  let context: number[] = new Array(attention.config.valueDim).fill(0);
  let confidence = 0;
  let tick = 0;

  for (tick = 0; tick < maxTicks; tick++) {
    // 1. Attend
    context = attention.forward(syncMatrix, embeddings);

    // 2. Update neurons
    updateNeurons(neuronActivations, context);

    // 3. Recompute sync matrix
    syncMatrix = computeSyncMatrix(neuronActivations);

    // 4. Check confidence
    confidence = confHead.forward(syncMatrix);
    if (confidence > threshold) {
      tick++; // count this tick
      break;
    }
  }

  const converged = confidence > threshold;
  const log: TickLoopLog = {
    ticksRun: tick,
    maxTicks,
    converged,
    confidence,
    confidenceThreshold: threshold,
  };

  return { context, confidence, ticksRun: tick, converged, log };
}
